File: Installer/Examples/Templates/ProjectTemplates/CSharp_Project/FileManager.cs
/*
 * FileManager.cs
 * Утилиты для работы с файловой системой на C#
 * Демонстрация для File Manager Pro
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace FileManagerPro.Examples.CSharpProject
{
    /// <summary>
    /// Типы файлов для классификации
    /// </summary>
    public enum FileType
    {
        Unknown,
        SourceCode,
        Script,
        Document,
        Data,
        Image,
        Archive,
        Executable
    }

    /// <summary>
    /// Информация о файле
    /// </summary>
    public class FileInfo
    {
        public string Name { get; set; }
        public long Size { get; set; }
        public string Extension { get; set; }
        public DateTime ModifiedDate { get; set; }
        public FileType Type { get; set; }

        public FileInfo(string name = "", long size = 0, string extension = "", DateTime? modifiedDate = null)
        {
            Name = name;
            Size = size;
            Extension = extension;
            ModifiedDate = modifiedDate ?? DateTime.Now;
            Type = DetermineFileType(extension);
        }

        private FileType DetermineFileType(string extension)
        {
            if (string.IsNullOrEmpty(extension))
                return FileType.Unknown;

            string ext = extension.ToLower().TrimStart('.');

            return ext switch
            {
                // Код
                "cs" or "cpp" or "c" or "h" or "hpp" or "java" => FileType.SourceCode,
                
                // Скрипты
                "py" or "js" or "ts" or "php" or "rb" or "sh" => FileType.Script,
                
                // Документы
                "txt" or "md" or "rtf" or "doc" or "docx" or "pdf" => FileType.Document,
                
                // Данные
                "json" or "xml" or "yaml" or "yml" or "csv" or "ini" => FileType.Data,
                
                // Изображения
                "jpg" or "jpeg" or "png" or "gif" or "bmp" or "svg" => FileType.Image,
                
                // Архивы
                "zip" or "rar" or "7z" or "tar" or "gz" => FileType.Archive,
                
                // Исполняемые файлы
                "exe" or "dll" or "so" or "dylib" => FileType.Executable,
                
                _ => FileType.Unknown
            };
        }
    }

    /// <summary>
    /// Утилиты для работы с файловой системой
    /// </summary>
    public static class FileManager
    {
        /// <summary>
        /// Проверяет валидность имени файла
        /// </summary>
        public static bool IsValidFilename(string filename)
        {
            if (string.IsNullOrWhiteSpace(filename) || filename.Length > 255)
                return false;

            // Запрещенные символы в Windows
            char[] invalidChars = Path.GetInvalidFileNameChars();
            
            if (filename.Any(c => invalidChars.Contains(c)))
                return false;

            // Запрещенные имена в Windows
            string[] reservedNames = {
                "CON", "PRN", "AUX", "NUL",
                "COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8", "COM9",
                "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9"
            };

            string nameWithoutExt = Path.GetFileNameWithoutExtension(filename).ToUpper();
            return !reservedNames.Contains(nameWithoutExt);
        }

        /// <summary>
        /// Получает расширение файла без точки
        /// </summary>
        public static string GetFileExtension(string filepath)
        {
            string ext = Path.GetExtension(filepath);
            return string.IsNullOrEmpty(ext) ? "" : ext.TrimStart('.');
        }

        /// <summary>
        /// Форматирует размер файла в читаемом виде
        /// </summary>
        public static string FormatFileSize(long bytes)
        {
            if (bytes == 0) return "0 Б";

            string[] units = { "Б", "КБ", "МБ", "ГБ", "ТБ" };
            int unitIndex = 0;
            double size = bytes;

            while (size >= 1024 && unitIndex < units.Length - 1)
            {
                size /= 1024;
                unitIndex++;
            }

            return unitIndex == 0 
                ? $"{size:0} {units[unitIndex]}" 
                : $"{size:0.##} {units[unitIndex]}";
        }

        /// <summary>
        /// Проверяет, является ли файл частью C# проекта
        /// </summary>
        public static bool IsCSharpProjectFile(string filepath)
        {
            string ext = GetFileExtension(filepath).ToLower();
            string[] csharpExtensions = { "cs", "csproj", "sln", "xaml", "config" };
            return csharpExtensions.Contains(ext);
        }

        /// <summary>
        /// Получает относительный путь
        /// </summary>
        public static string GetRelativePath(string fullPath, string basePath)
        {
            try
            {
                Uri fullUri = new Uri(fullPath);
                Uri baseUri = new Uri(basePath);

                return Uri.UnescapeDataString(
                    baseUri.MakeRelativeUri(fullUri).ToString()
                ).Replace('/', Path.DirectorySeparatorChar);
            }
            catch
            {
                return fullPath;
            }
        }

        /// <summary>
        /// Сканирует виртуальную структуру проекта (без реальных файлов)
        /// </summary>
        public static List<FileInfo> ScanVirtualProjectStructure()
        {
            return new List<FileInfo>
            {
                new FileInfo("Program.cs", 2048, ".cs", DateTime.Now.AddDays(-1)),
                new FileInfo("FileManager.cs", 4096, ".cs", DateTime.Now),
                new FileInfo("Utils.cs", 1024, ".cs", DateTime.Now.AddDays(-2)),
                new FileInfo("appsettings.json", 512, ".json", DateTime.Now.AddDays(-3)),
                new FileInfo("README.md", 256, ".md", DateTime.Now.AddDays(-4)),
                new FileInfo("project.csproj", 1024, ".csproj", DateTime.Now.AddDays(-5)),
                new FileInfo("config.xml", 768, ".xml", DateTime.Now.AddDays(-6))
            };
        }

        /// <summary>
        /// Анализирует структуру проекта
        /// </summary>
        public static string AnalyzeProjectStructure()
        {
            var files = ScanVirtualProjectStructure();
            
            var sb = new StringBuilder();
            sb.AppendLine("=== Анализ проекта C# ===");
            sb.AppendLine($"Всего файлов: {files.Count}");
            sb.AppendLine($"Общий размер: {FormatFileSize(files.Sum(f => f.Size))}");
            
            sb.AppendLine("\nПо типам файлов:");
            var byType = files.GroupBy(f => f.Type)
                             .OrderByDescending(g => g.Count());
            
            foreach (var group in byType)
            {
                sb.AppendLine($"  {group.Key}: {group.Count()} файлов");
            }
            
            sb.AppendLine("\nСамые большие файлы:");
            var largestFiles = files.OrderByDescending(f => f.Size).Take(3);
            foreach (var file in largestFiles)
            {
                sb.AppendLine($"  {file.Name}: {FormatFileSize(file.Size)}");
            }
            
            sb.AppendLine("========================");
            return sb.ToString();
        }
    }

    /// <summary>
    /// Класс для демонстрации работы с проектами
    /// </summary>
    public class ProjectAnalyzer
    {
        public string ProjectName { get; }
        public List<FileInfo> Files { get; }
        public List<string> Folders { get; }
        
        public ProjectAnalyzer(string projectName)
        {
            ProjectName = projectName;
            Files = new List<FileInfo>();
            Folders = new List<string>();
        }
        
        public void AddFile(FileInfo file)
        {
            Files.Add(file);
        }
        
        public void AddFolder(string folderName)
        {
            Folders.Add(folderName);
        }
        
        public string GetProjectReport()
        {
            var sb = new StringBuilder();
            
            sb.AppendLine($"Отчет по проекту: {ProjectName}");
            sb.AppendLine($"Дата создания: {DateTime.Now:yyyy-MM-dd HH:mm:ss}");
            sb.AppendLine();
            
            sb.AppendLine("Структура проекта:");
            sb.AppendLine($"  Папок: {Folders.Count}");
            sb.AppendLine($"  Файлов: {Files.Count}");
            sb.AppendLine($"  Общий размер: {FileManager.FormatFileSize(Files.Sum(f => f.Size))}");
            
            if (Files.Any())
            {
                sb.AppendLine("\nСтатистика по типам файлов:");
                var typeGroups = Files.GroupBy(f => f.Type)
                                     .OrderByDescending(g => g.Sum(f => f.Size));
                
                foreach (var group in typeGroups)
                {
                    sb.AppendLine($"  {group.Key}: {group.Count()} файлов, " +
                                 $"{FileManager.FormatFileSize(group.Sum(f => f.Size))}");
                }
            }
            
            return sb.ToString();
        }
    }

    /// <summary>
    /// Пример использования File Manager Pro с C#
    /// </summary>
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("========================================");
            Console.WriteLine("File Manager Pro - Демонстрация C#");
            Console.WriteLine("========================================\n");
            
            // Демонстрация валидации имен файлов
            Console.WriteLine("1. Валидация имен файлов:");
            Console.WriteLine($"   'Program.cs': {FileManager.IsValidFilename("Program.cs")}");
            Console.WriteLine($"   'file<bad>.txt': {FileManager.IsValidFilename("file<bad>.txt")}");
            Console.WriteLine($"   'CON.exe': {FileManager.IsValidFilename("CON.exe")}");
            
            // Демонстрация работы с расширениями
            Console.WriteLine("\n2. Работа с расширениями:");
            Console.WriteLine($"   Расширение 'utils.cs': {FileManager.GetFileExtension("utils.cs")}");
            Console.WriteLine($"   Это C# файл: {FileManager.IsCSharpProjectFile("utils.cs")}");
            Console.WriteLine($"   Это C# файл: {FileManager.IsCSharpProjectFile("data.json")}");
            
            // Демонстрация форматирования размеров
            Console.WriteLine("\n3. Форматирование размеров:");
            Console.WriteLine($"   1024 байт: {FileManager.FormatFileSize(1024)}");
            Console.WriteLine($"   1048576 байт: {FileManager.FormatFileSize(1048576)}");
            Console.WriteLine($"   1073741824 байт: {FileManager.FormatFileSize(1073741824)}");
            
            // Анализ виртуального проекта
            Console.WriteLine("\n4. Анализ проекта:");
            Console.WriteLine(FileManager.AnalyzeProjectStructure());
            
            // Создание и анализ проекта
            Console.WriteLine("\n5. Создание проекта:");
            var analyzer = new ProjectAnalyzer("FileManagerPro Demo");
            
            analyzer.AddFolder("src");
            analyzer.AddFolder("docs");
            analyzer.AddFolder("tests");
            
            analyzer.AddFile(new FileInfo("Program.cs", 2048, ".cs"));
            analyzer.AddFile(new FileInfo("Utils.cs", 1024, ".cs"));
            analyzer.AddFile(new FileInfo("README.md", 256, ".md"));
            analyzer.AddFile(new FileInfo("config.json", 512, ".json"));
            
            Console.WriteLine(analyzer.GetProjectReport());
            
            Console.WriteLine("\n========================================");
            Console.WriteLine("File Manager Pro поддерживает C# проекты:");
            Console.WriteLine("• Подсветка синтаксиса C#");
            Console.WriteLine("• Навигация по классам и методам");
            Console.WriteLine("• Поиск по всему проекту");
            Console.WriteLine("• Интеграция с .NET CLI");
            Console.WriteLine("========================================\n");
            
            // Демонстрация относительных путей
            if (args.Length > 0)
            {
                string fullPath = Path.GetFullPath(args[0]);
                string currentDir = Directory.GetCurrentDirectory();
                string relativePath = FileManager.GetRelativePath(fullPath, currentDir);
                
                Console.WriteLine($"Относительный путь: {relativePath}");
            }
        }
    }
}

File: Installer/Examples/Templates/ProjectTemplates/CSharp_Project/Cpp_Project/src/FileManager.cpp
/*
 * FileManager.cpp
 * Чистые утилиты для работы с файловыми путями
 * Без генерации файлов и побочных эффектов
 */

#include "FileManager.hpp"
#include <algorithm>
#include <cctype>
#include <sstream>

namespace fs = std::filesystem;

namespace FileManager {

// ============================================
// Реализация функций FileManager
// ============================================

std::vector<fs::path> listFiles(const fs::path& directory,
                                const std::string& filter) {
    std::vector<fs::path> result;
    
    // В реальном проекте здесь был бы код для сканирования директории
    // Для демонстрации возвращаем пустой список
    
    std::cout << "[FileManager] Запрос списка файлов для: " << directory << std::endl;
    std::cout << "[FileManager] Фильтр: " << (filter.empty() ? "*" : filter) << std::endl;
    
    // Возвращаем пустой список, так как это демонстрация без реальных файлов
    return result;
}

bool isCppProjectFile(const fs::path& filepath) {
    if (!filepath.has_extension()) {
        return false;
    }
    
    std::string ext = filepath.extension().string();
    std::transform(ext.begin(), ext.end(), ext.begin(),
                  [](unsigned char c) { return std::tolower(c); });
    
    const std::vector<std::string> cppExtensions = {
        ".cpp", ".cc", ".cxx", ".c++",
        ".hpp", ".hh", ".hxx", ".h++",
        ".h", ".c",
        ".ipp", ".inl", ".tpp", ".txx"
    };
    
    return std::find(cppExtensions.begin(), cppExtensions.end(), ext) != cppExtensions.end();
}

bool isValidFilename(const std::string& filename) {
    if (filename.empty() || filename.length() > 255) {
        return false;
    }
    
    // Запрещенные символы в именах файлов Windows
    const std::string invalidChars = "<>:\"/\\|?*";
    
    for (char c : filename) {
        // Проверка на управляющие символы (0-31)
        if (c < 32) {
            return false;
        }
        
        // Проверка на запрещенные символы
        if (invalidChars.find(c) != std::string::npos) {
            return false;
        }
    }
    
    // Запрещенные имена в Windows
    const std::vector<std::string> reservedNames = {
        "CON", "PRN", "AUX", "NUL",
        "COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8", "COM9",
        "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9"
    };
    
    std::string upperName = filename;
    std::transform(upperName.begin(), upperName.end(), upperName.begin(),
                  [](unsigned char c) { return std::toupper(c); });
    
    // Удаляем расширение для проверки
    size_t dotPos = upperName.find_last_of('.');
    std::string nameWithoutExt = (dotPos != std::string::npos) ? 
                                  upperName.substr(0, dotPos) : upperName;
    
    if (std::find(reservedNames.begin(), reservedNames.end(), nameWithoutExt) != reservedNames.end()) {
        return false;
    }
    
    return true;
}

fs::path normalizePath(const fs::path& path) {
    if (path.empty()) {
        return path;
    }
    
    try {
        // Приводим к абсолютному пути и нормализуем
        fs::path normalized = fs::absolute(path).lexically_normal();
        
        // Убираем trailing slash для директорий
        std::string pathStr = normalized.string();
        if (pathStr.length() > 1 && 
            (pathStr.back() == '/' || pathStr.back() == '\\')) {
            pathStr.pop_back();
        }
        
        return fs::path(pathStr);
    } catch (const fs::filesystem_error&) {
        // В случае ошибки возвращаем оригинальный путь
        return path;
    }
}

std::string getFileExtension(const fs::path& filepath) {
    if (!filepath.has_extension()) {
        return "";
    }
    
    std::string ext = filepath.extension().string();
    
    // Убираем точку в начале
    if (!ext.empty() && ext[0] == '.') {
        ext = ext.substr(1);
    }
    
    // Приводим к нижнему регистру
    std::transform(ext.begin(), ext.end(), ext.begin(),
                  [](unsigned char c) { return std::tolower(c); });
    
    return ext;
}

FileType getFileType(const fs::path& filepath) {
    if (!filepath.has_extension()) {
        return FileType::UNKNOWN;
    }
    
    std::string ext = getFileExtension(filepath);
    
    // Код
    if (ext == "cpp" || ext == "cc" || ext == "cxx" || 
        ext == "hpp" || ext == "hh" || ext == "hxx" || 
        ext == "c" || ext == "h") {
        return FileType::SOURCE_CODE;
    }
    
    // Скрипты
    if (ext == "py" || ext == "js" || ext == "ts" || 
        ext == "java" || ext == "cs" || ext == "php") {
        return FileType::SCRIPT;
    }
    
    // Документы
    if (ext == "txt" || ext == "md" || ext == "rtf" || 
        ext == "doc" || ext == "docx" || ext == "pdf") {
        return FileType::DOCUMENT;
    }
    
    // Данные
    if (ext == "json" || ext == "xml" || ext == "yaml" || 
        ext == "yml" || ext == "csv" || ext == "ini") {
        return FileType::DATA;
    }
    
    // Изображения
    if (ext == "jpg" || ext == "jpeg" || ext == "png" || 
        ext == "gif" || ext == "bmp" || ext == "svg") {
        return FileType::IMAGE;
    }
    
    // Архивы
    if (ext == "zip" || ext == "rar" || ext == "7z" || 
        ext == "tar" || ext == "gz") {
        return FileType::ARCHIVE;
    }
    
    return FileType::UNKNOWN;
}

std::string formatFileSize(uintmax_t bytes) {
    if (bytes == 0) return "0 Б";
    
    const char* units[] = {"Б", "КБ", "МБ", "ГБ", "ТБ"};
    int unitIndex = 0;
    double size = static_cast<double>(bytes);
    
    while (size >= 1024.0 && unitIndex < 4) {
        size /= 1024.0;
        unitIndex++;
    }
    
    std::ostringstream oss;
    if (unitIndex == 0) {
        oss << static_cast<int>(size) << " " << units[unitIndex];
    } else {
        oss << std::fixed << std::setprecision(2) << size << " " << units[unitIndex];
    }
    
    return oss.str();
}

fs::path combinePaths(const fs::path& base, const fs::path& relative) {
    if (relative.empty()) return base;
    
    // Если относительный путь абсолютный, возвращаем его
    if (relative.is_absolute()) {
        return relative;
    }
    
    // Комбинируем пути
    try {
        return (base / relative).lexically_normal();
    } catch (const fs::filesystem_error&) {
        return base / relative;
    }
}

bool isPathWithinDirectory(const fs::path& path, const fs::path& directory) {
    try {
        // Приводим пути к нормальной форме
        fs::path normPath = fs::absolute(path).lexically_normal();
        fs::path normDir = fs::absolute(directory).lexically_normal();
        
        // Проверяем, что путь начинается с директории
        auto mismatch = std::mismatch(normDir.begin(), normDir.end(),
                                      normPath.begin());
        
        return mismatch.first == normDir.end();
    } catch (const fs::filesystem_error&) {
        return false;
    }
}

std::string getRelativePath(const fs::path& path, const fs::path& base) {
    try {
        fs::path absolutePath = fs::absolute(path);
        fs::path absoluteBase = fs::absolute(base);
        
        // Пытаемся получить относительный путь
        fs::path relative = fs::relative(absolutePath, absoluteBase);
        
        if (relative.empty()) {
            return ".";
        }
        
        return relative.string();
    } catch (const fs::filesystem_error&) {
        return path.string();
    }
}

void demonstratePathManipulation() {
    std::cout << "\n=== Манипуляции с путями (FileManager) ===" << std::endl;
    
    // Примеры путей
    fs::path examplePath = "C:/Projects/FileManagerPro/src/main.cpp";
    fs::path baseDir = "C:/Projects/FileManagerPro";
    
    std::cout << "Пример пути: " << examplePath << std::endl;
    std::cout << "Имя файла: " << examplePath.filename() << std::endl;
    std::cout << "Родительская папка: " << examplePath.parent_path() << std::endl;
    std::cout << "Расширение: " << getFileExtension(examplePath) << std::endl;
    std::cout << "Тип файла: " << static_cast<int>(getFileType(examplePath)) << std::endl;
    
    // Комбинирование путей
    fs::path combined = combinePaths(baseDir, "build/output.exe");
    std::cout << "\nКомбинированный путь: " << combined << std::endl;
    
    // Относительный путь
    std::string relative = getRelativePath(examplePath, baseDir);
    std::cout << "Относительный путь: " << relative << std::endl;
    
    // Валидация имен файлов
    std::cout << "\nВалидация имен файлов:" << std::endl;
    std::cout << "  'main.cpp': " << (isValidFilename("main.cpp") ? "✓" : "✗") << std::endl;
    std::cout << "  'file<bad>.txt': " << (isValidFilename("file<bad>.txt") ? "✓" : "✗") << std::endl;
    std::cout << "  'CON.txt': " << (isValidFilename("CON.txt") ? "✓" : "✗") << std::endl;
    
    // Проверка C++ файлов
    std::cout << "\nПроверка C++ файлов:" << std::endl;
    std::cout << "  'utils.hpp': " << (isCppProjectFile("utils.hpp") ? "✓" : "✗") << std::endl;
    std::cout << "  'data.json': " << (isCppProjectFile("data.json") ? "✓" : "✗") << std::endl;
}

} // namespace FileManager